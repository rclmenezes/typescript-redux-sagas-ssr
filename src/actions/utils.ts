export type Error<T> = { [P in keyof T]?: string };

export type FormikSetErrors<T> = (errors: Error<T>) => void;

export interface FormikPayload<D> {
  data: D;
  setErrors: FormikSetErrors<D>;
}

export interface FormikTokenPayload<P> extends FormikPayload<P> {
  token: string;
}
