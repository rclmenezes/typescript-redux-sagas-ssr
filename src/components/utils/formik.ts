import { AxiosError } from "axios";

export type Error<T> = { [P in keyof T]?: string | undefined };

type FormikSetErrors<T> = (errors: T) => void;

export function setFormikErrorsFromError<T>(error: AxiosError, setErrors: FormikSetErrors<T>) {
  if (error.response && error.response.status === 400 && error.response.data) {
    setErrors(error.response.data);
  }
}
