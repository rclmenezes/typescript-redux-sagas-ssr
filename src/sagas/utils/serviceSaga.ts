import axios, { AxiosError, AxiosInstance } from "axios";
import { SagaIterator } from "redux-saga";
import { call, put, takeLatest } from "redux-saga/effects";
import { Action, AsyncActionCreators } from "typescript-fsa";

import { IS_CLIENT, PORT } from "../../settings";

let axiosInstance: AxiosInstance;
if (IS_CLIENT) {
  axiosInstance = axios.create();
} else {
  axiosInstance = axios.create({
    baseURL: `http://localhost:${PORT}`,
  });
}

export type Get = "GET";
export type Post = "POST";
export type MethodTypes = Get | Post;
export const GET: Get = "GET";
export const POST: Post = "POST";

export function createService<Data, Return>(url: string, method: MethodTypes) {
  return async (data: Data, headers: Record<string, string> | null) => {
    const res = await axiosInstance.request({
      data,
      headers,
      method,
      url,
    });
    return res.data as Return;
  };
}

export function createServiceSaga<P, S, E, D>(options: {
  actionCreator: AsyncActionCreators<P, S, E>;
  extractDataFromAction?: (action: Action<P>) => D;
  extractHeadersFromAction?: ((action: Action<P>) => Record<string, string>) | null;
  onError?: (action: Action<P>, e: AxiosError) => SagaIterator;
  onSuccess?: (action: Action<P>, result: S) => SagaIterator;
  method?: MethodTypes;
  url: string;
}) {
  const defaultOptions = {
    extractDataFromAction: (action: Action<P>) => (action.payload as unknown) as D,
    extractHeadersFromAction: null,
    method: GET,
  };
  const {
    actionCreator,
    extractDataFromAction,
    extractHeadersFromAction,
    onError,
    onSuccess,
    method,
    url,
  } = {
    ...defaultOptions,
    ...options,
  };

  const service = createService(url, method);

  const saga = function*(action: Action<P>) {
    const data = extractDataFromAction(action);
    const headers = extractHeadersFromAction ? extractHeadersFromAction(action) : null;
    try {
      const result: S = yield call(service, data, headers);
      yield put(
        actionCreator.done(
          {
            params: action.payload,
            result,
          },
          action.meta,
        ),
      );
      if (onSuccess) {
        yield* onSuccess(action, result);
      }
    } catch (err) {
      yield put(
        actionCreator.failed(
          {
            error: err,
            params: action.payload,
          },
          action.meta,
        ),
      );
      if (onError) {
        yield* onError(action, err);
      }
    }
  };
  return function*() {
    yield takeLatest<Action<P>>(actionCreator.started.type, saga);
  };
}
