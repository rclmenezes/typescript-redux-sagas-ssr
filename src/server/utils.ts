import { NextFunction, Request, Response } from "express";
// import sequelize from "../db/sequelize";

// export function wrapAsync<T>(fn: (req: Request, res: Response) => Promise<T>) {
//   return (req: Request, res: Response, next: NextFunction) => {
//     sequelize
//       .transaction(t => {
//         return fn(req, res);
//       })
//       .catch(next);
//   };
// }

export function wrapAsync<T>(fn: (req: Request, res: Response, next: NextFunction) => Promise<T>) {
  return (req: Request, res: Response, next: NextFunction) => {
    Promise.resolve(fn(req, res, next)).catch(next);
  };
}
