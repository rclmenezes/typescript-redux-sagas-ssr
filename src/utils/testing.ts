import { Href, UnregisterCallback } from "history";

import { PageRoutingProps } from "./routing";

export function createMockRouterProps<P>(params: P, search: string = "") {
  const location = {
    hash: "",
    key: "",
    pathname: "",
    search,
    state: {},
  };

  const props: PageRoutingProps<P> = {
    history: {
      action: "POP",
      block: t => {
        const temp: UnregisterCallback = () => {};
        return temp;
      },
      createHref: t => {
        const temp: Href = "";
        return temp;
      },
      go: num => {},
      goBack: () => {},
      goForward: () => {},
      length: 2,
      listen: t => {
        const temp: UnregisterCallback = () => {};
        return temp;
      },
      location,
      push: () => {},
      replace: () => {},
    },
    location,
    match: {
      isExact: true,
      params,
      path: "",
      url: "",
    },
    staticContext: {
      status: 200,
    },
  };
  return props;
}
