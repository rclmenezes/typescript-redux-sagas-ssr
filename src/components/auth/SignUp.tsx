import * as csstips from "csstips";
import { Field, Form, FormikBag, FormikProps, withFormik } from "formik";
import * as React from "react";
import { Link } from "react-router-dom";
import { Button, Message } from "semantic-ui-react";
import { style } from "typestyle";
import * as Yup from "yup";

import { SignUp, SignUpPayload } from "../../actions/auth";
import { PageRoutingProps } from "../../utils/routing";
import InputField from "../common/Input";
import BaseAuthPage from "./BaseAuthPage";

export interface SignUpProps {
  routing: PageRoutingProps<{}>;
  signUpSubmit: (payload: SignUpPayload) => void;
}

const InnerSignUpForm: React.SFC<FormikProps<SignUp> & SignUpProps> = () => (
  <Form>
    <Field name="email" component={InputField} placeholder="Email" fluid />
    <Field name="password" component={InputField} placeholder="Password" type="password" fluid />
    <Button type="submit" fluid primary>
      Submit
    </Button>
    <Message className={style(csstips.horizontallyCenterChildren)}>
      <Link to="/login">Already have an account?</Link>
    </Message>
  </Form>
);

const SignUpForm = withFormik<SignUpProps, SignUp>({
  handleSubmit: (values: SignUp, formikBag: FormikBag<SignUpProps, SignUp>) => {
    formikBag.props.signUpSubmit({
      data: values,
      setErrors: formikBag.setErrors,
    });
  },
  mapPropsToValues: () => {
    return {
      email: "",
      password: "",
    };
  },
  validationSchema: Yup.object().shape({
    email: Yup.string().required("Email is a required field"),
    password: Yup.string()
      .required("Password is a required field")
      .min(8, "Password must be at least 8 characters long"),
  }),
})(InnerSignUpForm);

class SignUpPage extends React.Component<SignUpProps> {
  public render() {
    return (
      <BaseAuthPage title="Sign Up">
        <SignUpForm {...this.props} />
      </BaseAuthPage>
    );
  }
}

export default SignUpPage;
