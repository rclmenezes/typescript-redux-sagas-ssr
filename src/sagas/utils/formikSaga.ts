import { AxiosError } from "axios";
import { call } from "redux-saga/effects";
import { Action, AsyncActionCreators } from "typescript-fsa";

import { FormikPayload } from "../../actions/utils";
import { createServiceSaga, POST } from "./serviceSaga";

export { FormikPayload };

export type FormikAction<P> = Action<FormikPayload<P>>;
export type FormikAsyncActionCreator<P, S, E> = AsyncActionCreators<FormikPayload<P>, S, E>;

export function* handleFormErrors<P>(action: FormikAction<P>, error: AxiosError) {
  if (error.response && error.response.status === 400 && error.response.data) {
    yield call(action.payload.setErrors, error.response.data);
  }
}

export function createFormikSubmitSaga<P, S, E>(
  actionCreator: FormikAsyncActionCreator<P, S, E>,
  url: string,
  extractHeadersFromAction: ((action: FormikAction<P>) => {}) | null = null,
) {
  return createServiceSaga({
    actionCreator,
    extractDataFromAction: action => action.payload.data,
    extractHeadersFromAction,
    method: POST,
    onError: handleFormErrors,
    url,
  });
}
