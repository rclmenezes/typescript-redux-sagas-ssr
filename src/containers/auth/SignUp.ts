import { connect } from "react-redux";
import { bindActionCreators, Dispatch } from "redux";

import actionCreators from "../../actions";
import SignUp from "../../components/auth/SignUp";
import { RootState } from "../../reducers";

const mapStateToProps = (state: RootState) => ({
  user: state.user,
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      signUpSubmit: actionCreators.signUp.started,
    },
    dispatch,
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SignUp);
