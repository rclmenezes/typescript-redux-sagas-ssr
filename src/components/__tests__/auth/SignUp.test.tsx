import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";
import { MemoryRouter } from "react-router";

import { createMockRouterProps } from "../../../utils/testing";
import SignUp, { SignUpProps } from "../../auth/SignUp";

describe("Test SignUp", () => {
  const routing = createMockRouterProps({});
  const MOCK_SIGNUP = {
    email: "email@email.com",
    password: "12341234",
  };

  const setup = (propOverrides: Partial<SignUpProps> = {}) => {
    const signUpSubmit = jest.fn();

    const props = {
      routing,
      signUpSubmit,
      user: null,
      ...propOverrides,
    };

    const wrapper = mount(
      <MemoryRouter>
        <SignUp {...props} />
      </MemoryRouter>,
    );

    return {
      getEmailField: () => wrapper.find('FieldInner[name="email"]'),
      getFormProps: () => wrapper.find("InnerSignUpForm").props() as any,
      getPasswordField: () => wrapper.find('FieldInner[name="password"]'),
      props,
      setEmailInput: (email: string) =>
        wrapper.find('input[name="email"]').simulate("change", {
          persist: () => {},
          target: {
            name: "email",
            value: email,
          },
        }),
      setPasswordInput: (password: string) =>
        wrapper.find('input[name="password"]').simulate("change", {
          persist: () => {},
          target: {
            name: "password",
            value: password,
          },
        }),
      signUpSubmit,
      wrapper,
    };
  };

  test("Submit successfully", async () => {
    const { getFormProps, setEmailInput, signUpSubmit, setPasswordInput, wrapper } = setup();
    expect(toJson(wrapper.find(SignUp) as any)).toMatchSnapshot();

    setEmailInput(MOCK_SIGNUP.email);
    setPasswordInput(MOCK_SIGNUP.password);

    let formikProps = getFormProps();
    await formikProps.submitForm();
    wrapper.update();
    formikProps = getFormProps();

    expect(formikProps.isSubmitting).toBe(true);
    expect(formikProps.isValid).toBe(true);
    expect(signUpSubmit.mock.calls.length).toBe(1);

    expect(signUpSubmit.mock.calls[0][0].data).toEqual(MOCK_SIGNUP);
  });

  test("setErrors gets called", async () => {
    const {
      getFormProps,
      getPasswordField,
      setEmailInput,
      signUpSubmit,
      setPasswordInput,
      wrapper,
    } = setup();

    setEmailInput(MOCK_SIGNUP.email);
    setPasswordInput(MOCK_SIGNUP.password);

    let formikProps = getFormProps();
    await formikProps.submitForm();
    wrapper.update();
    formikProps = getFormProps();

    expect(formikProps.isSubmitting).toBe(true);
    expect(formikProps.isValid).toBe(true);
    expect(signUpSubmit.mock.calls.length).toBe(1);

    expect(signUpSubmit.mock.calls[0][0].data).toEqual(MOCK_SIGNUP);

    const passwordError = "badness";
    signUpSubmit.mock.calls[0][0].setErrors({ password: passwordError });
    expect(getPasswordField().text()).toBe(passwordError);
  });

  test("Test submit without input", async () => {
    const { getFormProps, getEmailField, getPasswordField, wrapper } = setup();

    let formikProps = getFormProps();
    await formikProps.submitForm();
    wrapper.update();

    formikProps = getFormProps();
    expect(formikProps.errors.email).toBeTruthy();
    expect(getEmailField().text()).toBe(formikProps.errors.email);
    expect(formikProps.errors.password).toBeTruthy();
    expect(getPasswordField().text()).toBe(formikProps.errors.password);
  });

  test("Test submit without password", async () => {
    const { setEmailInput, getFormProps, getPasswordField, wrapper } = setup();

    setEmailInput(MOCK_SIGNUP.email);

    let formikProps = getFormProps();
    await formikProps.submitForm();
    wrapper.update();

    formikProps = getFormProps();
    expect(formikProps.errors.password).toBeTruthy();
    expect(getPasswordField().text()).toBe(formikProps.errors.password);
  });

  test("Test submit with short password", async () => {
    const { setEmailInput, getFormProps, getPasswordField, setPasswordInput, wrapper } = setup();

    setEmailInput("test");
    setPasswordInput("1234567");

    let formikProps = getFormProps();
    await formikProps.submitForm();
    wrapper.update();

    formikProps = getFormProps();
    expect(formikProps.errors.password).toBeTruthy();
    expect(getPasswordField().text()).toBe(formikProps.errors.password);
  });
});
