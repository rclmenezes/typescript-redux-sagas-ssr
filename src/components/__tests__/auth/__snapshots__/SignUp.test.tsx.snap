// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SignUp Submit successfully 1`] = `
<SignUpPage
  routing={
    Object {
      "history": Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 2,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "",
          "pathname": "",
          "search": "",
          "state": Object {},
        },
        "push": [Function],
        "replace": [Function],
      },
      "location": Object {
        "hash": "",
        "key": "",
        "pathname": "",
        "search": "",
        "state": Object {},
      },
      "match": Object {
        "isExact": true,
        "params": Object {},
        "path": "",
        "url": "",
      },
      "staticContext": Object {
        "status": 200,
      },
    }
  }
  signUpSubmit={[MockFunction]}
  user={null}
>
  <BaseAuthPage
    title="Sign Up"
  >
    <Container
      text={true}
    >
      <div
        className="ui text container"
      >
        <Header
          as="h1"
          textAlign="center"
        >
          <h1
            className="ui center aligned header"
          >
            Sign Up
          </h1>
        </Header>
        <WithFormik(InnerSignUpForm)
          routing={
            Object {
              "history": Object {
                "action": "POP",
                "block": [Function],
                "createHref": [Function],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "length": 2,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "key": "",
                  "pathname": "",
                  "search": "",
                  "state": Object {},
                },
                "push": [Function],
                "replace": [Function],
              },
              "location": Object {
                "hash": "",
                "key": "",
                "pathname": "",
                "search": "",
                "state": Object {},
              },
              "match": Object {
                "isExact": true,
                "params": Object {},
                "path": "",
                "url": "",
              },
              "staticContext": Object {
                "status": 200,
              },
            }
          }
          signUpSubmit={[MockFunction]}
          user={null}
        >
          <Formik
            enableReinitialize={false}
            handleSubmit={[Function]}
            initialValues={
              Object {
                "email": "",
                "password": "",
              }
            }
            isInitialValid={false}
            onSubmit={[Function]}
            render={[Function]}
            routing={
              Object {
                "history": Object {
                  "action": "POP",
                  "block": [Function],
                  "createHref": [Function],
                  "go": [Function],
                  "goBack": [Function],
                  "goForward": [Function],
                  "length": 2,
                  "listen": [Function],
                  "location": Object {
                    "hash": "",
                    "key": "",
                    "pathname": "",
                    "search": "",
                    "state": Object {},
                  },
                  "push": [Function],
                  "replace": [Function],
                },
                "location": Object {
                  "hash": "",
                  "key": "",
                  "pathname": "",
                  "search": "",
                  "state": Object {},
                },
                "match": Object {
                  "isExact": true,
                  "params": Object {},
                  "path": "",
                  "url": "",
                },
                "staticContext": Object {
                  "status": 200,
                },
              }
            }
            signUpSubmit={[MockFunction]}
            user={null}
            validateOnBlur={true}
            validateOnChange={true}
            validationSchema={[Function]}
          >
            <InnerSignUpForm
              dirty={false}
              errors={Object {}}
              handleBlur={[Function]}
              handleChange={[Function]}
              handleReset={[Function]}
              handleSubmit={[Function]}
              initialValues={
                Object {
                  "email": "",
                  "password": "",
                }
              }
              isSubmitting={false}
              isValid={false}
              isValidating={false}
              registerField={[Function]}
              resetForm={[Function]}
              routing={
                Object {
                  "history": Object {
                    "action": "POP",
                    "block": [Function],
                    "createHref": [Function],
                    "go": [Function],
                    "goBack": [Function],
                    "goForward": [Function],
                    "length": 2,
                    "listen": [Function],
                    "location": Object {
                      "hash": "",
                      "key": "",
                      "pathname": "",
                      "search": "",
                      "state": Object {},
                    },
                    "push": [Function],
                    "replace": [Function],
                  },
                  "location": Object {
                    "hash": "",
                    "key": "",
                    "pathname": "",
                    "search": "",
                    "state": Object {},
                  },
                  "match": Object {
                    "isExact": true,
                    "params": Object {},
                    "path": "",
                    "url": "",
                  },
                  "staticContext": Object {
                    "status": 200,
                  },
                }
              }
              setError={[Function]}
              setErrors={[Function]}
              setFieldError={[Function]}
              setFieldTouched={[Function]}
              setFieldValue={[Function]}
              setFormikState={[Function]}
              setStatus={[Function]}
              setSubmitting={[Function]}
              setTouched={[Function]}
              setValues={[Function]}
              signUpSubmit={[MockFunction]}
              submitCount={0}
              submitForm={[Function]}
              touched={Object {}}
              unregisterField={[Function]}
              user={null}
              validateField={[Function]}
              validateForm={[Function]}
              validateOnBlur={true}
              validateOnChange={true}
              values={
                Object {
                  "email": "",
                  "password": "",
                }
              }
            >
              <Form>
                <Component
                  formik={
                    Object {
                      "dirty": false,
                      "errors": Object {},
                      "handleBlur": [Function],
                      "handleChange": [Function],
                      "handleReset": [Function],
                      "handleSubmit": [Function],
                      "initialValues": Object {
                        "email": "",
                        "password": "",
                      },
                      "isSubmitting": false,
                      "isValid": false,
                      "isValidating": false,
                      "registerField": [Function],
                      "resetForm": [Function],
                      "setError": [Function],
                      "setErrors": [Function],
                      "setFieldError": [Function],
                      "setFieldTouched": [Function],
                      "setFieldValue": [Function],
                      "setFormikState": [Function],
                      "setStatus": [Function],
                      "setSubmitting": [Function],
                      "setTouched": [Function],
                      "setValues": [Function],
                      "submitCount": 0,
                      "submitForm": [Function],
                      "touched": Object {},
                      "unregisterField": [Function],
                      "validate": undefined,
                      "validateField": [Function],
                      "validateForm": [Function],
                      "validateOnBlur": true,
                      "validateOnChange": true,
                      "validationSchema": [Function],
                      "values": Object {
                        "email": "",
                        "password": "",
                      },
                    }
                  }
                >
                  <form
                    onReset={[Function]}
                    onSubmit={[Function]}
                  >
                    <FormikConnect(FieldInner)
                      component={[Function]}
                      fluid={true}
                      name="email"
                      placeholder="Email"
                    >
                      <FieldInner
                        component={[Function]}
                        fluid={true}
                        formik={
                          Object {
                            "dirty": false,
                            "errors": Object {},
                            "handleBlur": [Function],
                            "handleChange": [Function],
                            "handleReset": [Function],
                            "handleSubmit": [Function],
                            "initialValues": Object {
                              "email": "",
                              "password": "",
                            },
                            "isSubmitting": false,
                            "isValid": false,
                            "isValidating": false,
                            "registerField": [Function],
                            "resetForm": [Function],
                            "setError": [Function],
                            "setErrors": [Function],
                            "setFieldError": [Function],
                            "setFieldTouched": [Function],
                            "setFieldValue": [Function],
                            "setFormikState": [Function],
                            "setStatus": [Function],
                            "setSubmitting": [Function],
                            "setTouched": [Function],
                            "setValues": [Function],
                            "submitCount": 0,
                            "submitForm": [Function],
                            "touched": Object {},
                            "unregisterField": [Function],
                            "validate": undefined,
                            "validateField": [Function],
                            "validateForm": [Function],
                            "validateOnBlur": true,
                            "validateOnChange": true,
                            "validationSchema": [Function],
                            "values": Object {
                              "email": "",
                              "password": "",
                            },
                          }
                        }
                        name="email"
                        placeholder="Email"
                      >
                        <InputField
                          field={
                            Object {
                              "name": "email",
                              "onBlur": [Function],
                              "onChange": [Function],
                              "value": "",
                            }
                          }
                          fluid={true}
                          form={
                            Object {
                              "dirty": false,
                              "errors": Object {},
                              "handleBlur": [Function],
                              "handleChange": [Function],
                              "handleReset": [Function],
                              "handleSubmit": [Function],
                              "initialValues": Object {
                                "email": "",
                                "password": "",
                              },
                              "isSubmitting": false,
                              "isValid": false,
                              "isValidating": false,
                              "registerField": [Function],
                              "resetForm": [Function],
                              "setError": [Function],
                              "setErrors": [Function],
                              "setFieldError": [Function],
                              "setFieldTouched": [Function],
                              "setFieldValue": [Function],
                              "setFormikState": [Function],
                              "setStatus": [Function],
                              "setSubmitting": [Function],
                              "setTouched": [Function],
                              "setValues": [Function],
                              "submitCount": 0,
                              "submitForm": [Function],
                              "touched": Object {},
                              "unregisterField": [Function],
                              "validateField": [Function],
                              "validateForm": [Function],
                              "validateOnBlur": true,
                              "validateOnChange": true,
                              "values": Object {
                                "email": "",
                                "password": "",
                              },
                            }
                          }
                          placeholder="Email"
                        >
                          <div
                            className="f1urtfz"
                          >
                            <Input
                              fluid={true}
                              name="email"
                              onBlur={[Function]}
                              onChange={[Function]}
                              placeholder="Email"
                              type="text"
                              value=""
                            >
                              <div
                                className="ui fluid input"
                              >
                                <input
                                  name="email"
                                  onBlur={[Function]}
                                  onChange={[Function]}
                                  placeholder="Email"
                                  type="text"
                                  value=""
                                />
                              </div>
                            </Input>
                          </div>
                        </InputField>
                      </FieldInner>
                    </FormikConnect(FieldInner)>
                    <FormikConnect(FieldInner)
                      component={[Function]}
                      fluid={true}
                      name="password"
                      placeholder="Password"
                      type="password"
                    >
                      <FieldInner
                        component={[Function]}
                        fluid={true}
                        formik={
                          Object {
                            "dirty": false,
                            "errors": Object {},
                            "handleBlur": [Function],
                            "handleChange": [Function],
                            "handleReset": [Function],
                            "handleSubmit": [Function],
                            "initialValues": Object {
                              "email": "",
                              "password": "",
                            },
                            "isSubmitting": false,
                            "isValid": false,
                            "isValidating": false,
                            "registerField": [Function],
                            "resetForm": [Function],
                            "setError": [Function],
                            "setErrors": [Function],
                            "setFieldError": [Function],
                            "setFieldTouched": [Function],
                            "setFieldValue": [Function],
                            "setFormikState": [Function],
                            "setStatus": [Function],
                            "setSubmitting": [Function],
                            "setTouched": [Function],
                            "setValues": [Function],
                            "submitCount": 0,
                            "submitForm": [Function],
                            "touched": Object {},
                            "unregisterField": [Function],
                            "validate": undefined,
                            "validateField": [Function],
                            "validateForm": [Function],
                            "validateOnBlur": true,
                            "validateOnChange": true,
                            "validationSchema": [Function],
                            "values": Object {
                              "email": "",
                              "password": "",
                            },
                          }
                        }
                        name="password"
                        placeholder="Password"
                        type="password"
                      >
                        <InputField
                          field={
                            Object {
                              "name": "password",
                              "onBlur": [Function],
                              "onChange": [Function],
                              "value": "",
                            }
                          }
                          fluid={true}
                          form={
                            Object {
                              "dirty": false,
                              "errors": Object {},
                              "handleBlur": [Function],
                              "handleChange": [Function],
                              "handleReset": [Function],
                              "handleSubmit": [Function],
                              "initialValues": Object {
                                "email": "",
                                "password": "",
                              },
                              "isSubmitting": false,
                              "isValid": false,
                              "isValidating": false,
                              "registerField": [Function],
                              "resetForm": [Function],
                              "setError": [Function],
                              "setErrors": [Function],
                              "setFieldError": [Function],
                              "setFieldTouched": [Function],
                              "setFieldValue": [Function],
                              "setFormikState": [Function],
                              "setStatus": [Function],
                              "setSubmitting": [Function],
                              "setTouched": [Function],
                              "setValues": [Function],
                              "submitCount": 0,
                              "submitForm": [Function],
                              "touched": Object {},
                              "unregisterField": [Function],
                              "validateField": [Function],
                              "validateForm": [Function],
                              "validateOnBlur": true,
                              "validateOnChange": true,
                              "values": Object {
                                "email": "",
                                "password": "",
                              },
                            }
                          }
                          placeholder="Password"
                          type="password"
                        >
                          <div
                            className="f1urtfz"
                          >
                            <Input
                              fluid={true}
                              name="password"
                              onBlur={[Function]}
                              onChange={[Function]}
                              placeholder="Password"
                              type="password"
                              value=""
                            >
                              <div
                                className="ui fluid input"
                              >
                                <input
                                  name="password"
                                  onBlur={[Function]}
                                  onChange={[Function]}
                                  placeholder="Password"
                                  type="password"
                                  value=""
                                />
                              </div>
                            </Input>
                          </div>
                        </InputField>
                      </FieldInner>
                    </FormikConnect(FieldInner)>
                    <Button
                      as="button"
                      fluid={true}
                      primary={true}
                      type="submit"
                    >
                      <button
                        className="ui fluid primary button"
                        onClick={[Function]}
                        type="submit"
                      >
                        Submit
                      </button>
                    </Button>
                    <Message
                      className="fsuivmx"
                    >
                      <div
                        className="ui message fsuivmx"
                      >
                        <Link
                          replace={false}
                          to="/login"
                        >
                          <a
                            href="/login"
                            onClick={[Function]}
                          >
                            Already have an account?
                          </a>
                        </Link>
                      </div>
                    </Message>
                  </form>
                </Component>
              </Form>
            </InnerSignUpForm>
          </Formik>
        </WithFormik(InnerSignUpForm)>
      </div>
    </Container>
  </BaseAuthPage>
</SignUpPage>
`;
